### PR TITLE
Create a Repository wrapper to existing APIs/models

### PR DESCRIPTION
**Pull request by rain-on**
This provides a fluent API to perfomr CRUD operations on:
* Spaces
* Environments
* Releases
* Projects
* ProjectGroups

This will need to be extended as use cases present.

### CODE REVIEW

File: acceptance-tests/src/test/java/com/octopus/sdk/test/CreateReleaseAcceptanceTest.java [hunk 1]
```diff
@@ -39,18 +37,16 @@
 
   @BeforeEach
   public void createDeploymentAcceptanceTestSetup() throws IOException {
```
Comment (long-sean) [inline acceptance-tests/src/test/java/com/octopus/sdk/test/CreateReleaseAcceptanceTest.java:39]: Copy paste method name?

File: acceptance-tests/src/test/java/com/octopus/sdk/test/ProjectsAcceptanceTest.java [hunk 2]
```diff
@@ -29,18 +29,16 @@
...
   @Test
   public void returnsAProjectApiForAKnownSpace() {
-    final ProjectApi projectApi = ProjectApi.create(client, spaceHome);
-    assertThat(projectApi).isNotNull();
+    final ProjectApi projectRepository = createdSpace.projects();
+    assertThat(projectRepository).isNotNull();
   }
 
   @Test
   public void canCreateProjectWithinASpaceQueryForItAndDeleteIt() throws IOException {
```
Comment (long-sean) [inline acceptance-tests/src/test/java/com/octopus/sdk/test/ProjectsAcceptanceTest.java:37]: Test method name implies a delete action and respective assertion

File: octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java [hunk 3]
```diff
@@ -43,15 +44,16 @@ public BaseNamedResourceApi(
     super(client, rootPath, resourceType, collectionType);
   }
 
-  public List<RESPONSE_TYPE> getByPartialName(final String partialName) throws IOException {
+  public List<WRAPPED_TYPE> getByPartialName(final String partialName) throws IOException {
     Preconditions.checkNotNull(partialName, "Cannot search for a project with a null partial name");
```
Comment (long-sean) [inline octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java:48]: Error message needs to be generic?

File: octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java [hunk 4]
```diff
@@ -43,15 +44,16 @@ public BaseNamedResourceApi(
...
     return getByQuery(singletonMap("partialName", singletonList(partialName)));
   }
 
-  public Optional<RESPONSE_TYPE> getByName(final String completeName) throws IOException {
+  public Optional<WRAPPED_TYPE> getByName(final String completeName) throws IOException {
     Preconditions.checkNotNull(completeName, "Cannot search for a space with a null name");
```
Comment (long-sean) [inline octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java:53]: Error message needs to be generic

File: octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java [hunk 5]
```diff
@@ -61,7 +63,7 @@ public BaseNamedResourceApi(
     if (exactNameMatch.size() == 0) {
       return Optional.empty();
     } else if (exactNameMatch.size() == 1) {
-      return Optional.of(exactNameMatch.get(0));
+      return Optional.of(createServerObject(exactNameMatch.get(0)));
     } else {
       throw new IllegalStateException(
           "Octopus Server reports more than 1 space with an identical name");
```
Comment (long-sean) [inline octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java:69]: Error message needs to be generic?

**Review by long-sean — COMMENTED** [review]
Just some minor changes with naming

File: acceptance-tests/src/test/java/com/octopus/sdk/test/CreateReleaseAcceptanceTest.java (hunk 1 above)
Comment (rain-on) [reply acceptance-tests/src/test/java/com/octopus/sdk/test/CreateReleaseAcceptanceTest.java:39]: done

File: acceptance-tests/src/test/java/com/octopus/sdk/test/ProjectsAcceptanceTest.java (hunk 2 above)
Comment (rain-on) [reply acceptance-tests/src/test/java/com/octopus/sdk/test/ProjectsAcceptanceTest.java:37]: updated test with new operations and assertions

File: octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java (hunk 3 above)
Comment (rain-on) [reply octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java:48]: done

File: octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java (hunk 4 above)
Comment (rain-on) [reply octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java:53]: done

File: octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java (hunk 5 above)
Comment (rain-on) [reply octopus-sdk/src/main/java/com/octopus/sdk/api/BaseNamedResourceApi.java]: done

**Review by long-sean — APPROVED** [review]
Good to go

**Pull request merged**